lib/text/text.tsx

```diff
@@ -0,0 +1,317 @@
+        success: false
+      });
+
+      // ! Not ideal.
+      this.forceUpdate();
+    }
+
+    if (this.props.value !== previousProps.value) {
+      this.setState({
+        value: calculateValue(this.props.value, this.props.maxLength)
+      });
+    }
+  }
+
+  handleFocus = (event: React.FocusEvent): void => {
+    const { onFocus } = this.props;
+    const target: HTMLInputElement = event.target as HTMLInputElement;
+
+    // * select the input's value
+    target.select();
+
+    // * If the input type is password, then set the keyStrokes state to:
+    // * false.
+    if (this.props.type === `password`) {
+      this.setState({ keyStrokes: false });
+    }
+
+    onFocus(event);
+  };
+
+  handleKeyDown = (event: React.KeyboardEvent): void => {
+    // * event.keycode conflates keys on mobile keypad.
+    const key: string = event.key;
+
+    // * If the input type is password, there have been no previous key strokes
+    // * since focus and the backspace or delete keys are pressed, then set the
+    // * value state to: ''.
+    if (
+      this.props.type === `password` &&
+      !this.state.keyStrokes &&
+      (key === `Backspace` || key === `Delete`)
+    ) {
+      this.setState({ value: `` });
+    }
+
+    // * If the input type is number, then prevent any other keys from
+    // * leaking into the input's value.
+    if (this.props.type === `number`) {
+      const validKeys = [
+        `Backspace`,
+        `Tab`,
+        `ArrowLeft`,
+        `ArrowUp`,
+        `ArrowRight`,
+        `ArrowDown`,
+        `Delete`,
+        `0`,
+        `1`,
+        `2`,
+        `3`,
+        `4`,
+        `5`,
+        `6`,
+        `7`,
+        `8`,
+        `9`,
+        // * Mathematical operators
+        `-`,
+        `+`,
+        // * Allows selct all functionality.
+        `a`,
+        // * Allows copy, paste and cut functionality.
+        `c`,
+        `v`,
+        `x`
+      ];
+
+      if (validKeys.indexOf(key) === -1) {
+        event.preventDefault();
+      }
+    }
+  };
+
+  handleChange = (event: React.ChangeEvent) => {
+    const target: HTMLInputElement = event.target as HTMLInputElement;
+
+    this.setState({ value: target.value });
+
+    // * If the input type is password, then set the keyStrokes state to: true.
+    if (this.props.type === `password`) {
+      this.setState({ keyStrokes: true });
+    }
+
+    this.props.onChange(event, target.value, this.props.name);
+  };
+
+  render() {
+    const { props, state, id, handleFocus, handleKeyDown, handleChange } = this;
+
+    const modifiers = remove([`error`, `success`], props.modifiers);
+
+    const bem = BEM(`text`);
+    bem.addModifiers(modifiers);
+    bem.addModifiers(state.error ? `error` : ``);
+    bem.addModifiers(state.success ? `success` : ``);
+    bem.addClassNames(props.className);
+
+    return (
+      <div className={bem.getResult()} style={props.style}>
+        <input
+          id={id}
+          className={
```

Comment:
`${bem.getElement('input')}${state.value ? ' active' : ''}` ? Bit more concise?

Reply:
what state.value is? an object? if so, is the falsy value your'e treating _undefined_?
how about
state.value === undefined ? bem.getElement(`input`) : `${bem.getElement(`input`)} active`
or better
state.value !== undefined ? `${bem.getElement(`input`)} active` :  bem.getElement(`input`) ;

preferring positive conditions and having the regular flow as "if" and the exception as "else" are good practices

Reply:
But I like coercion `!!!state.value`...

Reply:
Seriously though, positive conditions sounds like a nice way to work.

Reply:
How about this?
```
<input
    className={`${bem.getElement(`input`)} ${
        typeof state.value !== `undefined` ? `active` : ``
    }`} 
/>
```

Reply:
Do this instead, and we have a deal (subtle difference that makes for cleaner output).
```<input
    ...
    className={`${bem.getElement(`input`)}${
        typeof state.value !== `undefined` ? ` active` : ``
    }`}
   ... 
/>```